### ISSUE TITLE
JIT: Invalid results/assertion errors with modulo ops

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/runtime#70333]
```csharp
// Generated by Fuzzlyn v1.5 on 2022-06-05 17:15:09
// Run on Arm64 MacOS
// Seed: 9776694006090703095
// Reduced from 187.4 KiB to 0.3 KiB in 00:00:51
// Debug: Outputs 255
// Release: Outputs 4294967295
public class Program
{
    public static void Main()
    {
        byte vr4 = default(byte);
        var vr5 = (byte)~vr4;
        uint vr6 = (uint)M50(vr5);
        System.Console.WriteLine(vr6);
    }

    public static long M50(byte arg0)
    {
        arg0 = 246;
        arg0 = (byte)(-1 % (arg0 | 1));
        return arg0;
    }
}
// Generated by Fuzzlyn v1.5 on 2022-06-05 14:45:52
// Run on Arm64 MacOS
// Seed: 18181223536262357280
// Reduced from 47.2 KiB to 0.3 KiB in 00:00:08
// Hits JIT assert in Release:
// Assertion failed 'FitsIn<int32_t>(value)' in 'Program:M1(ubyte)' during 'Morph - Global' (IL size 30; hash 0xb6989089; FullOpts)
// 
//     File: /Users/runner/work/1/s/src/coreclr/jit/compiler.hpp Line: 1517
// 
public class Program
{
    public static void Main()
    {
        M1(0);
    }

    public static void M1(byte arg0)
    {
        long var6 = default(long);
        arg0 = (byte)(~(ulong)var6 % 3545460779U);
        System.Console.WriteLine(arg0);
    }
}
```

**Comment by jakobbotsch** [comment]
@TIHan I wonder if this is fixed by #68885. It might also be exposed by some of the other recent PRs (I haven't seen these before).

cc @dotnet/jit-contrib

**Comment by TIHan** [comment]
Will take a look.

**Comment by TIHan** [comment]
Can confirm that these two are failing due to: https://github.com/dotnet/runtime/pull/69770

First one gives incorrect result.
Second one hits an assert.

Not immediately clear why, but looking into it.

**Comment by TIHan** [comment]
I sort of know what is happening. This PR https://github.com/dotnet/runtime/pull/69770 doesn't cause the issues here, but it does allow a casting bug to manifest itself.

Example:
(before morph)
```
               [000008] -A---------                         *  ASG       ubyte 
               [000007] D------N---                         +--*  LCL_VAR   ubyte  V00 arg0         
               [000006] -----------                         \--*  CAST      int <- ubyte <- long
               [000005] -----------                            \--*  UMOD      long  
               [000002] -----------                               +--*  NOT       long  
               [000001] -----------                               |  \--*  CNS_INT   long   0
               [000004] -----------                               \--*  CNS_INT   long   0xD353702B
```

(after morph) without [#69770](https://github.com/dotnet/runtime/pull/69770) change
```
               [000008] -A---+-----                         *  ASG       ubyte 
               [000007] D----+-N---                         +--*  LCL_VAR   ubyte  V00 arg0         
               [000021] -A---+--R--                         \--*  SUB       long  
               [000018] -----+-----                            +--*  LCL_VAR   long   V02 tmp1         
               [000020] -A---+-----                            \--*  MUL       long  
               [000005] MA---+-----                               +--*  UDIV      long  
               [000017] -A---+-----                               |  +--*  COMMA     long  
               [000015] -A---+-----                               |  |  +--*  ASG       long  
               [000014] D----+-N---                               |  |  |  +--*  LCL_VAR   long   V02 tmp1         
               [000002] -----+-----                               |  |  |  \--*  CNS_INT   long   -1
               [000016] -----+-----                               |  |  \--*  CNS_INT   long   -1
               [000004] -----+-N---                               |  \--*  CNS_INT   long   0xD353702B
               [000019] -----+-----                               \--*  CNS_INT   long   0xD353702B
```

(after morph) with [#69770](https://github.com/dotnet/runtime/pull/69770) change
```
               [000008] -A---+-----                         *  ASG       ubyte 
               [000007] D----+-N---                         +--*  LCL_VAR   ubyte  V00 arg0         
               [000021] -A---+-----                         \--*  COMMA     long  
               [000015] -A---+-----                            +--*  ASG       long  
               [000014] D----+-N---                            |  +--*  LCL_VAR   long   V02 tmp1         
               [000002] -----+-----                            |  \--*  CNS_INT   long   -1
               [000020] -----+-----                            \--*  CNS_INT   long   0xD2A5EF26
```

Looking above at both versions of the 'after morph', the cast has disappeared - this is because we allow narrowing at this point for `GT_ASG`:

`fgMorphSmpOp`
```cpp
            // If we are storing a small type, we might be able to omit a cast.
            if ((effectiveOp1->OperIs(GT_IND) ||
                 (effectiveOp1->OperIs(GT_LCL_VAR) &&
                  lvaGetDesc(effectiveOp1->AsLclVarCommon()->GetLclNum())->lvNormalizeOnLoad())) &&
                varTypeIsSmall(effectiveOp1))
            {
                if (!gtIsActiveCSE_Candidate(op2) && op2->OperIs(GT_CAST) &&
                    varTypeIsIntegral(op2->AsCast()->CastOp()) && !op2->gtOverflow())
                {
                    var_types castType = op2->CastToType();

                    // If we are performing a narrowing cast and
                    // castType is larger or the same as op1's type
                    // then we can discard the cast.

                    if (varTypeIsSmall(castType) && (genTypeSize(castType) >= genTypeSize(effectiveOp1)))
                    {
                        tree->AsOp()->gtOp2 = op2 = op2->AsCast()->CastOp();
                    }
                }
            }
``` 

I did experiments that fixed the issues here by either:
1. Removing this cast removal logic (regresses a lot of cases which makes sense)
2. Moving the cast removal to rationalization (few improvements and few regressions(more CSEs))

I still do not know the exact reason, but it does have something to do with the tree getting reduced more in global morph via assertion prop with the PR change.

**Comment by jakobbotsch** [comment]
What looks strange to me is removing an `int <- ... <- long` narrowing cast. It should be ok to remove casts on the form `int <- ... <- int` or `long <- ... <- long` for this case, but I'm not sure if the IR produced in your example (in both cases) is even valid.
What do the regressions look like if you add a further condition `genActualType(op2) == genActualType(op2->AsCast()->CastOp())`?

cc @dotnet/jit-contrib

**Comment by SingleAccretion** [comment]
There will be some regressions, I tried it in the past.

The immediate issue here is that we're making assertion for a `ubyte` (actually, `int`) variable with a `long` value, and then `optConstantAssertionProp` with `BashToConst` doesn't expect that kind of mismatch.

You would think you'd run into a similar issue with the `(int)LCL_VAR long -> LCL_VAR int` that morph does, but I don't you think you can because of the morphing order (we propagate in pre-order, while this narrowing is performed post-order), so I would be somewhat inclined to say that we should not allow `ASG(LCL_VAR ubyte, long)` in HIR.

**Comment by jakobbotsch** [comment]
> I would be somewhat inclined to say that we should not allow `ASG(LCL_VAR ubyte, long)` in HIR

I agree, that seems best to me too, I just hope the regressions for this case aren't too bad.

**Comment by jakobbotsch** [comment]
The latter example was fixed by #70518. The former one is a duplicate of #70607.
